jupyterhub/templates/ingress.yaml

```diff
@@ -14,6 +14,18 @@ metadata:
     {{- . | toYaml | trimSuffix "\n" | nindent 4 }}
   {{- end }}
 spec:
+  {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1/Ingress" }}
+  defaultBackend:
```

Comment:
https://kubernetes.io/docs/concepts/services-networking/ingress/#default-backend
> An Ingress with no rules sends all traffic to a single default backend. The defaultBackend is conventionally a configuration option of the Ingress controller and is not specified in your Ingress resources.
>
> If none of the hosts or paths match the HTTP request in the Ingress objects, the traffic is routed to your default backend.

Is this going to change the behaviour of k8s clusters running other applications?

Reply:
@MickeyShnaiderman-RecoLabs Following up from https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1314#issuecomment-776144297 could you please explain why you think this should be configured by Z2JH as opposed to in the ingress controller?

Reply:
I don't think it should. I misused the term "default backend" in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1314#issuecomment-776144297. I actually meant the Ingress controller itself, my mistake. What I really meant is that not specifying a host in Z2JH's values should just create a rule with no host, or, if the API version supports it, a wildcard host (i.e. `*`). This will cause the Ingress controller (mistakenly called "default backend") to accept all hosts on the given address to the created Ingress rule.

Reply:
See https://kubernetes.io/docs/concepts/services-networking/ingress/#:~:text=If%20you%20create%20an%20Ingress,based%20virtual%20host%20being%20required.

Reply:
__Somwhat off topic confusion form my end__

> or, if the API version supports it, a wildcard host (i.e. *).

I think v1beta1 didn't pin how the path specification should work, so different ingress controllers interpreted the same path rule differently. Hmmm, or rather, the default value in v1beta1 had a `pathType` of `ImplementationSpecific` which left it to the ingress controller to decide how to treat the path rule, while `v1` defaults `pathType` to `Prefix`.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#httpingresspath-v1-networking-k8s-io
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#httpingresspath-v1beta1-networking-k8s-io

---

> This will cause the Ingress controller (mistakenly called "default backend") to accept all hosts on __the given address__ to the created Ingress rule.

I'm a bit confused in general about having an ingress without a host. My mental model is that an ingress controller will proxy traffic based on multiple Ingress resources based on a host header. But, I guess it can be more vague than this, that for example an ingress controller in practice could allocate domain names to ingress resources without a domain name associated rule and then let all traffic to that specific domain name go there.

This touches the parts I don't know, and make me not confident about this PR in general. How will a default backend of an ingress typically influence ingress controllers? Is this declared in k8s specs or is it up to the ingress controller entirely without any guidance from k8s specifications?

I wonder for example, what happens if I have a nginx-ingress-controller and two separate Ingress resources with a default backend and no host associated rule - it won't know where to send the traffic arriving to the single IP allocated to it as an entrypoint to the k8s cluster.

Reply:
@MickeyShnaiderman-RecoLabs do you understand if there is a difference between omitting the `host` section, and declaring a `defaultBackend`?

Reply:
Oh!

The difference is that by omitting host in a rule, you can still specify a path, while default backend will target all hosts and all paths.

__Question:__ do we want a `defaultBackend` or a rule like beforebut where host is allowed to be omitted?

@MickeyShnaiderman-RecoLabs, do you explicitly need a `defaultBackend` Ingress rule, or would a rule without an explicit host do fine?

Reply:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingressrule-v1-networking-k8s-io

> If the host is unspecified, the Ingress routes all traffic based on the specified IngressRuleValue.

Reply:
@consideRatio, @manics: Thanks for finding this caveat! Didn't actually know that it is possible to make a default routing on a service. 

As for the difference, it seems you're right. As for @manics question, I think that **if** the Ingress controller doesn't allocate more than one external address for any of the Ingress objects in the cluster it **will** change the behaviour of the routing for the entire cluster. But, I guess that in my usage example from https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1314#issuecomment-776144297 the Ingress controller would get a new address allocated by AWS on every Ingress object, so specifying a `defaultBackend` or just omitting the host would have the same effect, especially when there is Z2JH's proxy Pod to handle different paths.

So yes, a rule without an explicit host would do just fine and possibly prevent problems in other people's clusters 😅